### ISSUE TITLE
Android TFlite inconsistent performance when app is not in focus.

### ISSUE THREAD

**Issue by Artaches** [tensorflow/tensorflow#35283]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Android 9.0 and 10.0
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: Pixel 3a (Android 10), tested on Android 9 as well with same bad behaviour
- TensorFlow installed from (source or binary):  tflite .so built from source
- TensorFlow version (use command below): 1.13, 1.14, 1.15, 2.0, 2.1
- Bazel version (if compiling from source): various; 0.21 to 0.29

Problem:
We have an audio processing application that runs in real time, including when the phone's screen is off. Using the built shared objects (tensorflowlite.so) from r1.13 and r1.14 branches, our processing time stays consistent. Locking the phone or minimizing the app does not affect the tflite inference times. Our app also uses a foreground service to make Android give us optimal process scheduling.

However, when using newer tflite shared objects (.so) built from r1.15, r2.0, and r2.1 the inference performance drops when the app is not in focus (either screen was locked, or our app minimized). The behaviour is especially bad when using more threads, e.g: `interpreter->SetNumThreads(2);`

**Describe the current behavior**
Android tflite model inference times are lower when app is not in focus with tflite versions r1.15, r2.0, r2.1. 
**Describe the expected behavior**
Exactly the same inference processing performance when the app is in focus or not.

Is there something in the API of newer tflite versions (r1.15 and newer) that I could play around with to fix this? Any help is greatly appreciated, thank you!

**Comment by jdduke** [comment]
We adopted a new matrix multiplication library in 1.15, which generally has offered significantly improved multi-threaded performance. Prior to 1.15, unless you explicitly pin your process to big cores, multi-threaded performance was extremely variable. In 1.15+, the performance is generally far more deterministic. However, we haven't done as much testing against scenarios where the screen is locked or the app loses focus.

If you're building from source, could you try adding `--define=tflite_with_ruy=false` to your bazel build command, and get back to us with the results? Thanks!

**Comment by Artaches** [comment]
This has resolved it, thank you for the tip! For our application we're only using 1-2 cores, so multi-threaded performance is not as critical as consistent inference times.

**Comment by jdduke** [comment]
> This has resolved it, thank you for the tip!

For completeness, how was it resolved? Using the `--define=tflite_with_ruy=false` flag? If so, would you mind sharing more about your model so we can investigate ruy-based performance? Thanks!

**Comment by Artaches** [comment]
Yes using the ``--define=tflite_with_ruy=false`` made the performance the same with screen off. We're using a LSTM based model that uses the `tf.nn.rnn_cell.LSTMCell` and `tf.nn.dynamic_rnn` classes.

**Comment by talumbau** [comment]
I'm glad this has resolved the issue! We are interested in reproducing on our side. Is it possible for you to upload the model?

The new matrix multiply library, ruy, shouldn't be invoked on the LSTM/RNN ops as of yet, so I'm interested in understanding why this fixed your issue.

Also, to clarify, is the following correct:

Old behavior: same inference latency whether or not the app is in focus

New (bad) behavior: inference time is *HIGHER* when the app is not in focus (i.e. the performance is not as good when the app is not in focus)

Thanks!

**Comment by bjacob** [comment]
Could you please try the following (With ruy. So in this experiment, do NOT build with `--define=tflite_with_ruy=false`):

Edit this line:
https://github.com/tensorflow/tensorflow/blob/e6983d538f140552d18fff6377a80c45236f1e36/tensorflow/lite/experimental/ruy/wait.cc#L66

increase the number of milliseconds here, change the current value 2 to something like 20 instead. does this help?

**Comment by Artaches** [comment]
@talumbau The behavior is indeed as you described; higher inference times when app is not in focus. I've gotten a model that I can share, could you email me @ artashes333@hotmail.com so I can share it?

@bjacob Changing the `DurationFromMilliseconds(N)` number has improved the performance when the app is not in focus by a bit. On Pixel 3a:
`DurationFromMilliseconds(2)`: ~3.3x increase in inference times
`DurationFromMilliseconds(20)`: ~1.6x increase in inference times

**Comment by bjacob** [comment]
Thanks for running this experiment. ~1.6x is sufficiently close to ~1x that we can draw the following conclusion as to what is the main phenomenon happening here --- there may well be additional things at play, but this should be the main thing, especially if you find that further increading this duration beyond 20ms further reduces the inference times.

On a mobile system like Android, the OS aggressively throttles the computing power available to applications in order to preserve battery charge. The throttling is even more aggressive for applications that aren't foreground. So as a first approximation you could say that everything here is "working as intended" from the perspective of the OS.

In multithreaded inference, TensorFlow Lite uses internal a pool of "worker" threads. After a thread is done with its current task, it has to wait for a next task to come up. The standard way of doing that, "condition-variable waiting", informs the OS about this thread being waiting. That's good because that allows to conserve battery charge. But Android is so aggressive at reducing CPU clock speeds and even turning CPU cores altogether, and then so conservative at ramping CPU cores up again, that if we enter that waiting state then there is a very large penalty for the next workload. It may take 100ms or more to reach back high clock speeds. That is why the above-linked wait.cc code uses some "spin-waiting", i.e. just running a loop until a certain number of milliseconds have elapsed, to avoid the worst of that issue.

The problem of course is that there is no ideal choice for this "spin duration", as it's a compromise between wasting battery charge and suffering very high latency penalties. The current 2ms value was empirically determined from a few benchmarks, but this depends not only on TFLite models (governing time gaps between matrix multiplications within the inference) but also on application-level usage patterns (governing time gaps between inferences). 2ms was a conservative choice to avoid wasting too much battery charge, but the gap between this 2ms on the one hand, and the 100ms order of magnitude of the penalty to ramp back up to high clock speed is so large, there will be lots of room for variation among applications as to what's the best choice here.

That is why I believe the best course of action here would be to make this time duration runtime-configurable:
1) in ruy, make this configurable, presumably by exposing a control on ruy::Context.
2) in TFLite, expose this through CpuBackendContext, then on the user-facing TFLite API (TfLiteInterpreter?)

I had filed an internal issue (b/135595069) - @jdduke @talumbau please consider staffing.

**Comment by talumbau** [comment]
I have some updates here. I ran our standard benchmark_model` tool on the model provided by Artaches (thanks!) and I am now confused by the results you report. The model is all in float and the performance is dominated by the FullyConnected op (it spends 99% of its time in FC). When run with standard options (i.e. with Ruy on) on a small core of a Pixel 4, you get:

... a snippet from the tool output:

```
Average inference timings in us: Warmup: 14022, Init: 1183, Inference: 13681.3

Profile (1 threads):

Thread 0 (1410 samples)

* 99.29% FullyConnected
  * 99.29% cpu_backend_gemm::Gemm
    * 99.29% cpu_backend_gemm::Gemm: CustomGemv

```

The profiling indicates that the model does not actually execute any code inside Ruy. To simplify things, I will just say that the CustomGemv code exists to handle GEMV operations because Ruy, being a general matrix multiplication library, does not do well at the GEMV case (this is changing, but not germane to this conversation).

When I rebuild the `benchmark_model` tool with `--define=tflite_with_ruy=false` and re-run, I get:

```
Average inference timings in us: Warmup: 66187.9, Init: 1295, Inference: 65619.1
                                                                                                                                                     
Profile (1 threads):
                                                                                                                                                     
Thread 0 (3577 samples)
                                                                                                                                                    
* 99.89% FullyConnected
  * 99.89% cpu_backend_gemm::Gemm
    * 99.89% cpu_backend_gemm::Gemm: general GEMM
```

so the inferemce time is approximately 5x worse.

So the things I don't understand would be:
1. How does changing a parameter of Ruy multi-threading impact the latency you see in your model, given that Ruy code is not invoked to run your model at this time?

2. How does adding threads make the inference time worse, since the benchmarking tool shows that, in the default case, the CustomGemV path takes advantage of threading and reduces latency?

I'm not sure how to proceed here. One thing: can you run your model on Pixel3a with the benchmark tool and verify some of the findings I am seeing? At least we would have some level of agreement on raw performance numbers (I have access to a Pixel 3 and Pixel 4, but not Pixel 3a)

**Comment by talumbau** [comment]
For completeness, I will mention that the output `cpu_backend_gemm::Gemm: general GEMM` in the second profiling indicates that another GEMM implementation was used. In this case, the GEMM operation used `eigen` since that is our fallback path is Ruy is not available.

**Comment by Artaches** [comment]
@talumbau There are two FC layers that the ``benchmark_model`` tool is showing. The first is actually an RNN LSTM cell (`tf.nn.rnn_cell.LSTMCell` + `tf.nn.static_rnn`; 5.5m params), the second is a small actual FC layer (`tf.keras.layers.Dense`; 327k params). The computation time of the FC layers make sense to me. 

Here are the numbers I get on the ``pixel 3a`` with the standard ``benchmark_model``:
1 thread: ``adb shell /data/local/tmp/benchmark_model --graph=/data/local/tmp/performance_issue_model.tflite --num_threads=1 --num_runs=5000``:
```
INFO: Initialized TensorFlow Lite runtime.
count=102 first=13137 curr=4872 min=4682 max=13137 avg=4913.32 std=828

Running benchmark for at least 5000 iterations and at least 1 seconds but terminate if exceeding 150 seconds.
count=5000 first=4883 curr=4832 min=4660 max=5049 avg=4833.36 std=53

[Overall] - Memory usage: max resident set size = 23.0469 MB, total malloc-ed size = 0.104813 MB
Average inference timings in us: Warmup: 4913.32, Init: 885, no stats: 4833.36
```

4 threads: ``adb shell /data/local/tmp/benchmark_model --graph=/data/local/tmp/performance_issue_model.tflite --num_threads=4 --num_runs=5000``:
```
INFO: Initialized TensorFlow Lite runtime.
count=140 first=15603 curr=3159 min=3056 max=15603 avg=3556.35 std=1387

Running benchmark for at least 5000 iterations and at least 1 seconds but terminate if exceeding 150 seconds.
count=5000 first=3340 curr=3175 min=3042 max=8959 avg=3199.51 std=277

[Overall] - Memory usage: max resident set size = 22.4844 MB, total malloc-ed size = 0.0189743 MB
Average inference timings in us: Warmup: 3556.35, Init: 849, no stats: 3199.51
```
   
` `
  
Now with a ``benchmark_model`` binary that was built with ``--define=tflite_with_ruy=false``:
1 thread: ``adb shell /data/local/tmp/benchmark_model_v21_ruyFalse --graph=/data/local/tmp/performance_issue_model.tflite --num_threads=1 --num_runs=5000``:
```
Running benchmark for at least 1 iterations and at least 0.5 seconds but terminate if exceeding 150 seconds.
count=162 first=10464 curr=3053 min=2900 max=10464 avg=3074.59 std=583

Running benchmark for at least 5000 iterations and at least 1 seconds but terminate if exceeding 150 seconds.
count=5000 first=3141 curr=3121 min=2893 max=3406 avg=3039.76 std=42

[Overall] - Memory usage: max resident set size = 23.0625 MB, total malloc-ed size = 0.104752 MB
Average inference timings in us: Warmup: 3074.59, Init: 1282, no stats: 3039.76
```
4 threads: ``adb shell /data/local/tmp/benchmark_model_v21_ruyFalse --graph=/data/local/tmp/performance_issue_model.tflite --num_threads=4 --num_runs=5000``:
```
INFO: Initialized TensorFlow Lite runtime.
count=161 first=10793 curr=3071 min=2949 max=10793 avg=3091.92 std=610

Running benchmark for at least 5000 iterations and at least 1 seconds but terminate if exceeding 150 seconds.
count=5000 first=3098 curr=3013 min=2898 max=3383 avg=3045.69 std=41

[Overall] - Memory usage: max resident set size = 23.0586 MB, total malloc-ed size = 0.104813 MB
Average inference timings in us: Warmup: 3091.92, Init: 997, no stats: 3045.6
```
I hope this helps!

**Comment by bjacob** [comment]
Regarding the previous 3 comments, the explanation is that tflite uses the ruy thread-pool directly in place like this CustomGemv, via the cpu_backend_threadpool facility:
https://github.com/tensorflow/tensorflow/blob/9c7d78bce12a6374748ce84dd9a4fc0a37606775/tensorflow/lite/kernels/cpu_backend_gemm_custom_gemv.h#L179

When tflite_with_ruy=true (the default on ARM), cpu_backend_threadpool maps onto the ruy thread-pool,
https://github.com/tensorflow/tensorflow/blob/9c7d78bce12a6374748ce84dd9a4fc0a37606775/tensorflow/lite/kernels/cpu_backend_threadpool.h#L32-L41

Which ends up in the wait.cc code that I linked above.

So as far as I can see, everything here is still perfectly consistent with the theory proposed above, and it would be very useful to invest in the above described project to expose direct control over the spin duration to the application, with a clear explanation of the compromises here, and some degree of acknowledgement that there is a fundamental conflict here between a mobile OS conserving battery charge, and allowing background activities to have as low latency as foreground activities.

**Comment by KevinCS17** [comment]
How to make the foreground service?..
After make the notification and sservice
What class need to be called to the foregroundservice.java class?

**Comment by Artaches** [comment]
We've ran into another weird performance issue with tflite (this time using tflite r2.4 runtime) on the Pixel 3a. We've discovered that when the app is out of focus (minimized or screen turned off), the Pixel 3a forces all the tflite threads to run on the very last core of the CPU. This is particularly bad when using more than 1 thread for tflite (`interpreter->SetNumThreads(2 or greater)`) since all the tflite threads compete for the same core, and the scheduling causes worse performance than just using 1 thread. This can be clearly seen using the Android Studio profiler (with NDK profiling setup and everything). I've attached a screenshot that demonstrates this. 
![image](https://user-images.githubusercontent.com/3443179/115820143-1adc8600-a3b5-11eb-9fac-b64f184fc084.png)


We've only seen this type of scheduling behavior on the Google Pixel 3a, many much worse phones (such as the Samsung Galaxy J3) don't do this. Can anything be done on developer's end? Is there any way in Android to encourage the device to not schedule all the tflite worker threads on the same core?

Shall I make a seperate issue for this?

**Comment by multiverse-tf** [comment]
> We've ran into another weird performance issue with tflite (this time using tflite r2.4 runtime) on the Pixel 3a. We've discovered that when the app is out of focus (minimized or screen turned off), the Pixel 3a forces all the tflite threads to run on the very last CPU core. This is particularly bad when using more than 1 thread for tflite (`interpreter->SetNumThreads(2 or greater)`) since all the tflite worker threads compete for the same core, and the scheduling causes worse performance than just using 1 thread. This can be clearly seen using the Android Studio profiler (with NDK profiling setup, since we're using C++ tflite APIs). I've attached a screenshot that demonstrates this scheduling behavior. The "App Start" mark is where tflite processing begins. Our application runs tflite processing about every 8ms, so it's crucial to get consistent performance.
> ![image](https://user-images.githubusercontent.com/3443179/115820143-1adc8600-a3b5-11eb-9fac-b64f184fc084.png)
> 

Many thanks for such a detailed perf. investigation! I think such an OS scheduling strategy on Pixel3a in this case isn't that surprising if considering the goal is to save power consumption, and the screen being off implies no user interaction w/ the phone?

> We've only seen this type of scheduling behavior on the Google Pixel 3a, many much worse phones (such as the Samsung Galaxy J3) don't do this. Can anything be done on developer's end? Is there any way in Android to encourage the device to not schedule all the tflite worker threads on the same core? Any help is greatly appreciated!

These are great questions! I don't think Android has official APIs to set cpu affinity for threads as it's not encouraged to do so. Even if there's sth. that could be done on Pixel 3a to mitigate this issue, considering each Android phone manufacturer may adopt different OS scheduling strategies for each phone, there might be no universally unified approach to fix this. I could be totally wrong here as I don't have much Android experience. Anyway, I'll keep this thread updated once I learned more about such a scheduling strategy on Pixel 3a.

> 
> Also, shall I make a separate issue for this?

**Comment by alexcohn** [comment]
> These are great questions! I don't think Android has official APIs to set cpu affinity for threads as it's not encouraged to do so. 

See a related discussion: https://github.com/tensorflow/tensorflow/issues/29910. Yes, it is possible to choose the fast cores (by parsing **/proc/cpuinfo**), and set the process's CPU affinity mask accordingly. The question whether this improves the overall performance of your application, can only be answered by experimenting.

**Comment by sachinprasadhs** [comment]
Could you please try in the latest Tensorflow version and report if you still find performance issue. Thanks!

**Comment by Artaches** [comment]
@sachinprasadhs I've re-ran profiling on the Google Pixel 3a (now using Android 12 SP2A.220505.008) using two diff Android tflite runtimes build from R2.10 and R2.11_RC1 branches.
Unfortunately I'm still seeing the same weird behavior of first fast core (CPU6) not getting anything scheduled on it when screen is turned off. No difference between R2.10 and R2.11rc1 runtimes. CPU6 doesn't even seem to downclock when screen is off to save power, despite nothing being scheduled on it... 
![image](https://user-images.githubusercontent.com/3443179/197170065-6bb0f261-f376-4c95-80bc-60f60847c2a6.png)

**Comment by grantjensen** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The TFLite team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the TFLite space.
